src/document/doc-utils.js

```diff
@@ -182,17 +182,31 @@ function removeLines(doc) {
   return mapDoc(doc, removeLinesFn);
 }
 
+function getInnerParts(doc) {
+  let { parts } = doc;
+  let lastPart;
+  // Avoid a falsy element like ""
+  for (let i = 0; i < doc.parts.length && !lastPart; i++) {
+    lastPart = parts[parts.length - (i + 1)];
+  }
+  if (lastPart.type === "group") {
+    parts = lastPart.contents.parts;
+  }
+  return parts;
+}
+
 function stripTrailingHardline(doc) {
   // HACK remove ending hardline, original PR: #1984
   if (doc.type === "concat" && doc.parts.length !== 0) {
-    const lastPart = doc.parts[doc.parts.length - 1];
+    const parts = getInnerParts(doc);
+    const lastPart = parts[parts.length - 1];
     if (lastPart.type === "concat") {
       if (
         lastPart.parts.length === 2 &&
         lastPart.parts[0].hard &&
         lastPart.parts[1].type === "break-parent"
       ) {
-        return { type: "concat", parts: doc.parts.slice(0, -1) };
+        return { type: "concat", parts: parts.slice(0, -1) };
```

Comment:
Potential breaking change for plugins

Reply:
I think we can implement the `withInnerParts` argument (and maybe put todo to change it on default for prettier@3)

/cc @thorn0

Reply:
Thanks. I didn't notice. This change is needed for applying `stripTrailingHardline` to doc printed via `printer-yaml`. I'll try to fix `printer-yaml`.

Reply:
Are `doc-utils` considered part of the public plugin API?